pkg/controllers/namespaceregistration/controller.go

```diff
@@ -114,19 +122,100 @@ func (c *Controller) handleDelete(ctx context.Context, namespaceRegistration *ls
+
+	logger, ctx := logging.FromContextOrNew(ctx, nil)
+
+	// check if installations, executions, deploy items or target sync objects are still there
+	installations := &v1alpha1.InstallationList{}
+	if err := c.Client().List(ctx, installations, client.InNamespace(namespaceRegistration.GetName())); err != nil {
+		return c.logAndUpdate(ctx, err, namespaceRegistration, "Failed Reading Installations")
+	}
+
+	if len(installations.Items) > 0 {
+		for i := range installations.Items {
+			nextInst := &installations.Items[i]
+
+			// delete root installations with delete without uninstall annotation
+			if !utils.HasLabel(&nextInst.ObjectMeta, v1alpha1.EncompassedByLabel) && utils.HasDeleteWithoutUninstallAnnotation(&nextInst.ObjectMeta) {
+				if nextInst.GetDeletionTimestamp().IsZero() {
+					if err := c.Client().Delete(ctx, nextInst); err != nil {
+						logger.Error(err, "failed deleting installations without uninstall: "+client.ObjectKeyFromObject(nextInst).String())
+					}
+				} else if nextInst.Status.JobID == nextInst.Status.JobIDFinished && !helper.HasOperation(nextInst.ObjectMeta, v1alpha1.ReconcileOperation) {
+					// retrigger
+					metav1.SetMetaDataAnnotation(&nextInst.ObjectMeta, v1alpha1.OperationAnnotation, string(v1alpha1.ReconcileOperation))
+					if err := c.Client().Update(ctx, nextInst); err != nil {
+						logger.Error(err, "failed annotating installations without uninstall: "+client.ObjectKeyFromObject(nextInst).String())
+					}
+				}
+			}
+		}
+
+		return c.logAndUpdate(ctx, nil, namespaceRegistration, "Namespace Contains Installations")
+	}
+
+	executions := &v1alpha1.ExecutionList{}
+	if err := c.Client().List(ctx, executions, client.InNamespace(namespaceRegistration.GetName())); err != nil {
+		return c.logAndUpdate(ctx, err, namespaceRegistration, "Failed Reading Executions")
+	}
+
+	if len(executions.Items) > 0 {
+		return c.logAndUpdate(ctx, nil, namespaceRegistration, "Namespace Contains Executions")
+	}
+
+	deployItems := &v1alpha1.DeployItemList{}
+	if err := c.Client().List(ctx, deployItems, client.InNamespace(namespaceRegistration.GetName())); err != nil {
+		return c.logAndUpdate(ctx, err, namespaceRegistration, "Failed Reading DeployItems")
+	}
+
+	if len(deployItems.Items) > 0 {
+		return c.logAndUpdate(ctx, nil, namespaceRegistration, "Namespace Contains DeployItems")
+	}
+
+	targetSyncs := &v1alpha1.TargetSyncList{}
+	if err := c.Client().List(ctx, targetSyncs, client.InNamespace(namespaceRegistration.GetName())); err != nil {
+		return c.logAndUpdate(ctx, err, namespaceRegistration, "Failed Reading TargetSyncs")
+	}
+
+	if len(targetSyncs.Items) > 0 {
+		for i := range targetSyncs.Items {
+			nextTargetSync := &targetSyncs.Items[i]
+			controllerutil.RemoveFinalizer(nextTargetSync, v1alpha1.LandscaperFinalizer)
+
+			if err := c.Client().Update(ctx, nextTargetSync); err != nil {
+				return c.logAndUpdate(ctx, err, namespaceRegistration, "Failed Removing Finalizer Of TargetSync")
+			}
```

Comment:
Is it necessary to remove the finalizer? The TargetSync controller would directly add it again, if it is fast enough. 

Can we just delete the TargetSync (without removing the finalizer), and return as in the other cases where objects still exist. The TargetSync controller would then perform the deletion in the regular way.

Reply:
very good point

Reply:
I fixed it